src/server/pfs/server/obj_block_api_server.go

```diff
@@ -889,19 +889,78 @@ func (s *objBlockAPIServer) GetObjDirect(request *pfsclient.GetObjDirectRequest,
 	defer func(start time.Time) { s.Log(request, nil, retErr, time.Since(start)) }(time.Now())
 	r, err := s.objClient.Reader(server.Context(), request.Obj, 0, 0)
 	if err != nil {
-		return err
-	}
-	if err := grpcutil.WriteToStreamingBytesServer(r, server); err != nil {
+		if s.isNotFoundErr(err) {
+			return errors.Errorf("object %s not found with direct access", request.Obj)
+		}
 		return err
 	}
 	defer func() {
 		if err := r.Close(); err != nil && retErr == nil {
 			retErr = err
 		}
 	}()
+	if err := grpcutil.WriteToStreamingBytesServer(r, server); err != nil {
+		return err
+	}
 	return nil
 }
 
+func (s *objBlockAPIServer) DeleteObjDirect(ctx context.Context, request *pfsclient.DeleteObjDirectRequest) (response *types.Empty, retErr error) {
+	func() { s.Log(request, nil, nil, 0) }()
+	defer func(start time.Time) { s.Log(request, nil, retErr, time.Since(start)) }(time.Now())
+	if (request.Object != "") == (request.Prefix != "") {
+		return nil, errors.New("must specify either 'obj' or 'prefix' to delete")
+	}
+
+	if request.Prefix != "" {
+		cleanPrefix := path.Clean(request.Prefix)
+
+		// Try to avoid footguns that would delete everything
+		if cleanPrefix == "." ||
+			strings.HasPrefix(cleanPrefix, s.blockDir()) ||
+			strings.HasPrefix(cleanPrefix, s.objectDir()) ||
+			strings.HasPrefix(cleanPrefix, s.tagDir()) ||
+			strings.HasPrefix(cleanPrefix, s.indexDir()) {
+			return nil, errors.New("prefix-based direct object deletion is not allowed on system paths")
+		}
+
+		eg, ctx := errgroup.WithContext(ctx)
+		deleteChan := make(chan string, 50)
+		eg.Go(func() error {
+			limiter := limit.New(20)
+			for obj := range deleteChan {
+				object := obj
+				limiter.Acquire()
+				eg.Go(func() error {
+					defer limiter.Release()
+					return s.objClient.Delete(ctx, object)
+				})
+			}
+			return nil
+		})
+
+		if err := s.objClient.Walk(ctx, request.Prefix, func(name string) error {
+			deleteChan <- name
+			return nil
```

Comment:
I think it's possible to simplify this code by directly launching the deletion goroutine in the walk function rather than using a channel. Not a big deal if you like the code the way it is now.

Reply:
herp derp yep